implode.go

```diff
@@ -0,0 +1,134 @@
+//
+// 2. If both arg1 and arg2 are provided:
+//   - If arg1 is an array, arg2 is converted to string and used as the separator.
+//   - If arg1 is not an array and arg2 is an array, arg1 is converted to a string
+//     and used as the separator, with arg2 being the array to implode.
+//   - If neither arg1 nor arg2 is an array, the function returns an error.
+//
+// Non-string elements within the array are converted to strings using a ConvertToString function
+// before joining.
+// Due to language differences between PHP and Go, the implode function support OrderedMap type from the [orderedmap library],
+// ensuring ordered map functionality. When imploding map types, please utilize the OrderedMap type from the [orderedmap library]
+// to maintain element order. If you use map type, not OrderedMap type, the order of the results cannot be guaranteed.
+//
+// reference:
+//   - implode: https://github.com/php/php-src/blob/php-5.6.40/ext/standard/string.c#L1229-L1269
+//   - php_implode: https://github.com/php/php-src/blob/php-5.6.40/ext/standard/string.c#L1141-L1224
+//
+// Test cases:
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/strings/implode.phpt
+//   - https://github.com/php/php-src/blob/php-5.6.40/ext/standard/tests/strings/implode1.phpt
+//
+// [official PHP documentation]: https://www.php.net/manual/en/function.implode.php
+// [orderedmap library]: https://pkg.go.dev/github.com/elliotchance/orderedmap/v2
+func Implode(arg1 any, options ...any) (string, error) {
+	var delim string
+	var arr []any
+
+	// Check arg1 is one of array, slice, map, or ordered ap
+	isArg1CollectionType := isCollectionType(arg1)
+
+	// Check if options is not provided
+	if len(options) == 0 {
+		if !isArg1CollectionType {
+			return "", fmt.Errorf("argument must be one of array, slice, or ordered map, but got %v", reflect.TypeOf(arg1))
+		}
+		aggregateValues(arg1, &arr)
+	} else {
+		arg2 := options[0]
+		// Check arg2 is one of array, slice, or ordered map
+		isArg2CollectionType := isCollectionType(arg2)
+
+		if isArg1CollectionType {
+			delim, _ = ConvertToString(arg2)
+			aggregateValues(arg1, &arr)
+		} else if !isArg1CollectionType && isArg2CollectionType {
+			delim, _ = ConvertToString(arg1)
+			aggregateValues(arg2, &arr)
+		} else {
+			return "", fmt.Errorf("invalid arguments passed, got %v, %v", reflect.TypeOf(arg1), reflect.TypeOf(arg2))
+		}
+	}
+
+	// Join arr elements with a delim
+	var builder strings.Builder
+	if len(arr) == 0 {
+		return "", nil
+	}
+	for i, item := range arr {
+		str, err := ConvertToString(item)
+
+		if err != nil {
+			return "", fmt.Errorf("unsupported type in array : %v", reflect.TypeOf(item))
+		} else {
+			builder.WriteString(str)
+		}
+		if i < len(arr)-1 {
+			builder.WriteString(delim)
+		}
+	}
+	return builder.String(), nil
+}
+
+// isOrderedMap checks if the argument is an instance of ordered map
+func isOrderedMap(arg any) bool {
+	_, ok := arg.(*orderedmap.OrderedMap[any, any])
```

Comment:
I think it would be better if it also covers for non-pointer values.

Reply:
refactored the function to support non-pointer value, as well. And added test case for this function to validate whether this function support both pointer and non-pointer